// @expo/config/build/ios/utils/Xcodeproj.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const xcode_1 = __importDefault(require("xcode"));
const pbxFile_1 = __importDefault(require("xcode/lib/pbxFile"));
const Paths = __importStar(require("../Paths"));
function getProjectName(projectRoot) {
    const sourceRoot = Paths.getSourceRoot(projectRoot);
    return path.basename(sourceRoot);
}
exports.getProjectName = getProjectName;
// TODO(brentvatne): I couldn't figure out how to do this with an existing
// higher level function exposed by the xcode library, but we should find out how to do
// that and replace this with it
function addFileToGroup(filepath, groupName, project) {
    const file = new pbxFile_1.default(filepath);
    file.uuid = project.generateUuid();
    file.fileRef = project.generateUuid();
    project.addToPbxFileReferenceSection(file);
    project.addToPbxBuildFileSection(file);
    project.addToPbxSourcesBuildPhase(file);
    const group = pbxGroupByPath(project, groupName);
    if (!group) {
        throw Error(`Group by name ${groupName} not found!`);
    }
    group.children.push({
        value: file.fileRef,
        comment: file.basename,
    });
    return project;
}
exports.addFileToGroup = addFileToGroup;
function splitPath(path) {
    // TODO: Should we account for other platforms that may not use `/`
    return path.split('/');
}
const findGroup = (group, name) => {
    if (!group) {
        return undefined;
    }
    return group.children.find(group => group.comment === name);
};
function findGroupInsideGroup(project, group, name) {
    var _a;
    const foundGroup = findGroup(group, name);
    if (foundGroup) {
        return (_a = project.getPBXGroupByKey(foundGroup.value)) !== null && _a !== void 0 ? _a : null;
    }
    return null;
}
function pbxGroupByPath(project, path) {
    const { firstProject } = project.getFirstProject();
    let group = project.getPBXGroupByKey(firstProject.mainGroup);
    const components = splitPath(path);
    for (const name of components) {
        const nextGroup = findGroupInsideGroup(project, group, name);
        if (nextGroup) {
            group = nextGroup;
        }
        else {
            return null;
        }
    }
    return group !== null && group !== void 0 ? group : null;
}
function ensureGroupRecursively(project, filepath) {
    const components = splitPath(filepath);
    const hasChild = (group, name) => group.children.find(({ comment }) => comment === name);
    const { firstProject } = project.getFirstProject();
    let topMostGroup = project.getPBXGroupByKey(firstProject.mainGroup);
    for (const pathComponent of components) {
        if (topMostGroup && !hasChild(topMostGroup, pathComponent)) {
            topMostGroup.children.push({
                comment: pathComponent,
                value: project.pbxCreateGroup(pathComponent, '""'),
            });
        }
        topMostGroup = project.pbxGroupByName(pathComponent);
    }
    return topMostGroup !== null && topMostGroup !== void 0 ? topMostGroup : null;
}
exports.ensureGroupRecursively = ensureGroupRecursively;
/**
 * Get the pbxproj for the given path
 */
function getPbxproj(projectRoot) {
    const projectPath = Paths.getPBXProjectPath(projectRoot);
    const project = xcode_1.default.project(projectPath);
    project.parseSync();
    return project;
}
exports.getPbxproj = getPbxproj;
/**
 * Get the productName for a project, if the name is using a variable `$(TARGET_NAME)`, then attempt to get the value of that variable.
 *
 * @param project
 */
function getProductName(project) {
    var _a, _b;
    let productName = project.productName;
    if (productName === '$(TARGET_NAME)') {
        const targetName = (_b = (_a = project.getFirstTarget()) === null || _a === void 0 ? void 0 : _a.firstTarget) === null || _b === void 0 ? void 0 : _b.productName;
        productName = targetName !== null && targetName !== void 0 ? targetName : project.productName;
    }
    return productName;
}
exports.getProductName = getProductName;
function getProjectSection(project) {
    return project.pbxProjectSection();
}
exports.getProjectSection = getProjectSection;
function getNativeTargets(project) {
    const section = project.pbxNativeTargetSection();
    return Object.entries(section).filter(isNotComment);
}
exports.getNativeTargets = getNativeTargets;
function findFirstNativeTarget(project) {
    const { targets } = Object.values(getProjectSection(project))[0];
    const target = targets[0].value;
    const nativeTargets = getNativeTargets(project);
    const nativeTarget = nativeTargets.find(([key]) => key === target)[1];
    return nativeTarget;
}
exports.findFirstNativeTarget = findFirstNativeTarget;
function getXCConfigurationListEntries(project) {
    const lists = project.pbxXCConfigurationList();
    return Object.entries(lists).filter(isNotComment);
}
exports.getXCConfigurationListEntries = getXCConfigurationListEntries;
function getBuildConfigurationForId(project, configurationListId) {
    const configurationListEntries = getXCConfigurationListEntries(project);
    const [, configurationList] = configurationListEntries.find(([key]) => key === configurationListId);
    const buildConfigurations = configurationList.buildConfigurations.map(i => i.value);
    return Object.entries(project.pbxXCBuildConfigurationSection())
        .filter(isNotComment)
        .filter(isBuildConfig)
        .filter(isNotTestHost)
        .filter(([key]) => buildConfigurations.includes(key));
}
exports.getBuildConfigurationForId = getBuildConfigurationForId;
function isBuildConfig([, sectionItem]) {
    return sectionItem.isa === 'XCBuildConfiguration';
}
exports.isBuildConfig = isBuildConfig;
function isNotTestHost([, sectionItem]) {
    return !sectionItem.buildSettings.TEST_HOST;
}
exports.isNotTestHost = isNotTestHost;
function isNotComment([key]) {
    return !key.endsWith(`_comment`);
}
exports.isNotComment = isNotComment;
